infrastructure/builder/src/build/monoimage.js

```diff
@@ -147,8 +156,54 @@ const generateMonoimageTasks = ({tasks, baseDir, spec, cfg, cmdOptions}) => {
     },
   });
 
-  // TODO: add hooks push for web-ui
-  // TODO: add entry point for docker run for web-ui
+  hooks.push({
+    name: 'web-ui',
+    build: async (requirements, utils, procs) => {
+      const cacheDir = path.join(workDir, 'ui/cache');
+      if (!fs.existsSync(cacheDir)) {
+        fs.mkdirSync(cacheDir);
+      }
+
+      utils.step({title: 'Run Yarn Install'});
+
+      await dockerRun({
+        image: nodeImage,
+        workingDir: '/app/ui',
+        env: ['YARN_CACHE_FOLDER=/ui/cache'],
+        command: ['yarn'],
+        logfile: `${workDir}/yarn.log`,
+        utils,
+        binds: [
+          `${appDir}:/app/ui`,
```

Comment:
This is mounting `appDir`, which is the repository root, at `/app/ui`, so the `yarn` command is running in the repository root -- and the next `dockerRun` thus runs `yarn build` in the root, which tries to use Docker, which doesn't work in a Docker container.  I think you'll want to mount `appDir` at `/app`, and rely on the `workingDir` definition above to make the `yarn` command run inside the `ui` subdirectory.  Same for the `dockerRun` below.

Reply:
What does "mounting" mean here? I was actually trying to find out what `binds` means - my understanding is this comes from `dockerode`, so I tried looking up docks in there and found no explanation, and then I tried to search in Docker docs for `docker run` comand - also didn't find anything

Reply:
It's equivalent to [-v](https://docs.docker.com/engine/reference/run/#volume-shared-filesystems).  I'm not sure why it's named "binds" instead of "volumes".  The format is "<dir-on-host>:<dir-in-container>" so `${appDir}:/app/ui` mounts whatever `appDir` points at, on the host filesystem, at `/app/ui` in the docker container.

Reply:
There are `volumes`: https://github.com/apocas/dockerode/blob/master/examples/external_volume.js#L10 `binds` appears to be separate

Reply:
Docker's pretty inconsistent in what it calls things.  Since dockerode is a wrapper around the Docker API, I start with the API docs.  So corresponding to the link in your comment, the createContainer docs are at https://docs.docker.com/engine/api/v1.37/#operation/ContainerCreate.  That has both Volumes and HostConfig.Binds.  I think Volumes just mounts empty temporary directories (that will be destroyed when the container exits) while Binds connects host directories to container directories:

> A list of volume bindings for this container. Each volume binding is a string in one of these forms:
> 
>     host-src:container-dest to bind-mount a host path into the container. Both host-src, and container-dest must be an absolute path.
>     host-src:container-dest:ro to make the bind mount read-only inside the container. Both host-src, and container-dest must be an absolute path.
>     volume-name:container-dest to bind-mount a volume managed by a volume driver into the container. container-dest must be an absolute path.
>     volume-name:container-dest:ro to mount the volume read-only inside the container. container-dest must be an absolute path.

Now that I look through that documentation, it looks like `HostConfig.Mounts` could do the same job.  Maybe that's a newer option or an older one?  Dunno, Docker is weird.

I don't specifically remember figuring this out, but my general sources for Docker learnings are:
 * The Docker API, because dockerode docs say "Input options are directly passed to Docker. Check Docker API documentation for more details."
 * Other things that use docker.  I vaguely recall looking at [docker-worker](https://github.com/taskcluster/docker-worker/blob/a930de7b6101f95caa1f91e986f40e27f3039abb/test/dockerworker.js#L96-L100)
 * Trial and error.  Docker API docs are only for the most recent version, which isn't necessarily what we're running, and sometimes things change incompatibly.

Reply:
oh, API docs... I see my mistake. I was looking at the CLI docs. I found `binds` in API docs. ty!